### ISSUE TITLE
Prevent autosaving while in the middle of an activity

### ISSUE THREAD

**Issue by AmarinReyny** [CleverRaven/Cataclysm-DDA#61762]
### Is your feature request related to a problem? Please describe.

I've experienced a lot of errors (too many to really go over, sorry) that seem to come after loading a save that was automatically made in the middle of an ongoing activity (especially when the activity is something like picking up/dropping/inserting/unloading items). I'm still trying to figure out what to do about the errors I'm getting, but in any case, the fact that these errors keep happening after autosaving mid-activity shows that there is a problem that should be solved so that they don't happen any more than they already have.

There's also the frustration of loading a save in the middle of sleeping, especially after I selected to save immediately before attempting to sleep in the first place - the autosave makes that option useless, since the save made before sleeping will be automatically overwritten without my input.

### Solution you would like.

I would like for there to be an option for turning off autosaving in the middle of activities. That, and maybe even a whitelist/blacklist feature that can allow the player to select which activities should prevent autosaving (like sleeping, moving items around/between containers/etc., consuming items, crafting, etc..).

### Describe alternatives you have considered.

Continuously receiving errors that cause an increasing amount of bugginess and performance loss as they accumulate while I try to figure out all over again how to fix them

### Additional context

I'm still trying to figure out what to do about the errors I'm getting now that are tied to autosaves made during activities - I vaguely remember having figured out the solution to them a long time ago, but I can't for the life of me recall what the solution was.

**Comment by mqrause** [comment]
What are the errors and what version are you playing? A bunch of activity saving errors have been fixed in the experimental version.

**Comment by AmarinReyny** [comment]
> What are the errors and what version are you playing? A bunch of activity saving errors have been fixed in the experimental version.

Mostly this one, though I suspect it's tied to multiple items, not just one: 

>  DEBUG    : item_location lost its target item during a save/load cycle

 FUNCTION : void item_location::impl::ensure_unpacked() const
 FILE     : src/item_location.cpp
 LINE     : 116
 VERSION  : 32ee0b1

**Comment by mqrause** [comment]
Which activity was that?

**Comment by mqrause** [comment]
Can you upload the save? Sounds like things got saved in a way they shouldn't have be saved in.

It's pretty uncritical, though, unless items actually got deleted. You should be able to play on after ignoring this particular kind of error.

**Comment by AmarinReyny** [comment]
> Can you upload the save? Sounds like things got saved in a way they shouldn't have be saved in.
> 
> It's pretty uncritical, though, unless items actually got deleted. You should be able to play on after ignoring this particular kind of error.

Do you only need the save, or will you also need the data the save uses? I use a lot of mods, but I've also been tweaking some of the base game files, as well - there are some changes I don't know how to put in a mod, but I can figure out how to make by changing already-existing data (such as basecamp recipes).

**Comment by AmarinReyny** [comment]
> Just the save should be enough (for now). Performance loss is more likely the NPCs fault, especially if there are multiple of them. You could test that by simply going far away enough.

All right. I'll work on uploading the save folder. It'll probably take a while, though - my wifi is shit, after all. In the meantime, I have already done numerous tests involving moving far away from where my base (and thus NPCs, pets, and hoarded items) is; nothing is restoring the performance of the game to what it was prior to the errors' occurance.

**Comment by AmarinReyny** [comment]
Well, that was quicker than I ever hoped it could be. Here's my save folder.

[https://drive.google.com/file/d/1AS5zkbkjYmcx4QknfM1_cHA017o0fIvv/view?usp=sharing](https://drive.google.com/file/d/1AS5zkbkjYmcx4QknfM1_cHA017o0fIvv/view?usp=sharing)

**Comment by mqrause** [comment]
Okay, so from what I can see, your player isn't in any activity. However you have 4 NPCs disassembling stuff and 3 of them want to disassemble the same item. Marion Lipscomb, Robbyn 'Snookie' Crane and Hildegarde Breaux all have the same activity target. I can't tell you what item they're aiming for, but stopping at least two of them should probably solve most of your issues.

**Comment by AmarinReyny** [comment]
> Okay, so from what I can see, your player isn't in any activity. However you have 4 NPCs disassembling stuff and 3 of them want to disassemble the same item. Marion Lipscomb, Robbyn 'Snookie' Crane and Hildegarde Breaux all have the same activity target. I can't tell you what item they're aiming for, but stopping at least two of them should probably solve most of your issues.
> 
> Edit: So this is most likely not a save/load issue. It's an NPC issue. And an issue of misleading error messages.

Interesting... I'll take two of them off disassembly duty, and see if that fixes it.

**Comment by mqrause** [comment]
No, with the current activity architecture, that's not something that could happen. At worst your activity gets saved wrong and is broken when you try to load. But as long as you still can load, the activity will just be cancelled or otherwise finished and that's it.

**Comment by AmarinReyny** [comment]
> No, with the current activity architecture, that's not something that could happen. At worst your activity gets saved wrong and is broken when you try to load. But as long as you still can load, the activity will just be cancelled or otherwise finished and that's it.

Ah, okay. Good to know, and thank you for clarifying that for me.

**Comment by AmarinReyny** [comment]
> No, with the current activity architecture, that's not something that could happen. At worst your activity gets saved wrong and is broken when you try to load. But as long as you still can load, the activity will just be cancelled or otherwise finished and that's it.

So... I just took everyone off disassembly duty, and waited for a while, just to see if anything would change. Everything is still extremely laggy. I guess the fact that the lag persisted even across Z-levels (which stop and prevent NPC activities such as disassembly) should have been a sign that stopping the disassembly wouldn't work...

**Comment by mqrause** [comment]
I recommend closing this and making a separate issue for the performance stuff where you attach the save and all the necessary mods to load it.

**Comment by AmarinReyny** [comment]
> My experience with zone activities across Z levels is that companions who have started an activity will finish it even if you change the Z level, but will be unable to continue on to the next activity since that zone isn't loaded. Setting companions to dig out a basement and then return to the surface to do something, such as e.g. read, will have the companions return up when done with the current segment. Thus, companions set to disassemble items ought to finish the current item, but then finish the disassembly activity as the zone isn't loaded.
> 
> A work around for auto save induced errors could be to disable auto save and rely on manual quicksaves instead. That obviously increases the manual activity burden a bit, though.

In my experience, NPCs will stop disassembling items almost immediately after changing Z-levels - I say "almost" because it usually happens the second after my survivor goes to a different Z-level. Granted, I also have NPCs chopping wood - one thing I've noticed is that NPCs who stopped disassembling will take some time before going back to it after my survivor returns to their Z-level, but the wood choppers will get right back to it the moment my survivor returns to their Z-level. Maybe it's different for them?

**Comment by AmarinReyny** [comment]
> I recommend closing this and making a separate issue for the performance stuff where you attach the save and all the necessary mods to load it.

All right. I'll get that uploaded when I can.
